exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -114,30 +118,82 @@ impl JavaRuntimeProxy {
 
                 ExceptionHandlers::DEFAULT(env, exception)
             }
-            _ => Error::OtherJniError.with_description(err.to_string()),
+            _ => Error::JniError.with_description(err.to_string()),
         }
     }
 
-    /// Executes closure `f` and handles any type of JNI errors from it.
+    /// Executes closure `f` and handles any type of JNI errors from it. Occurred
+    /// exceptions are cleared.
     ///
-    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
-    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
-    /// being thrown, `ExceptionHandlers::DEFAULT` is applied for such exceptions.
+    /// Used for runtime methods results of which do *not* depend on
+    /// implementation of the user service, but rather on implementation of
+    /// Java runtime and user input. E.g., `deploy_artifact`.
+    ///
+    /// - Any JNI errors are converted into `ExecutionError::Runtime(JniError)`.
+    /// - `IllegalArgumentException`s are converted into
+    ///   `ExecutionError::Runtime(IllegalArgument)`.
+    /// - Any other exceptions are converted into
+    ///   `ExecutionError::Runtime(JavaException)`.
     fn jni_call_default<F, R>(&self, f: F) -> Result<R, ExecutionError>
     where
         F: FnOnce(&JNIEnv) -> JniResult<R>,
     {
-        self.jni_call::<F, &ExceptionHandler, R>(&[], f)
+        self.jni_call::<F, &ExceptionHandler, R>(
+            &[(
+                &classes_refs::java_lang_illegal_argument_exception(),
+                ExceptionHandlers::ILLEGAL_ARGUMENT,
+            )],
+            f,
+        )
     }
 
-    /// Executes closure `f` and handles any type of JNI errors from it.
+    /// Executes closure `f` and handles any type of JNI errors from it. Occurred
+    /// exceptions are cleared.
     ///
-    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
-    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
-    /// being thrown, then exception is passed to corresponding `ExceptionHandler` according their
-    /// type and `exception_handlers` mapping.
-    /// `ExceptionHandlers::DEFAULT` is called in case of there is no any handlers or handlers are
-    /// not matched to exception type.
+    /// Used for runtime methods results of which depend on
+    /// implementation of the user service, such as `execute`,
+    /// `before_transactions` and `after_transactions`.
```

Comment:
I noticed 'CallSite' has 'constructor' variant:
// Service constructor.
CONSTRUCTOR = 1;

Shall we handle the errors from there in the same way?

Reply:
It's interesting. In what methods of the Runtime will the constructor be called? I assume, `initateAddingService`? We can cover this method too, if `TransactionExecutionException`s and/or `UnexpectedTransactionExecutionException`s are possible there.

Reply:
Currently, no, I submitted https://jira.bf.local/browse/ECR-4074 , but there are many questions.